### DISCUSSION TITLE
Why Preact?

### DISCUSSION THREAD

**Discussion by FeldrinH** [freshframework/fresh#703] [General]
I guess this is kind of a mean question, but I am qurious:
From what little I know about Preact, it is not particularly widely used and its market share doesn't seem to be growing either, so why use it as the basis of a new web framework?

**Comment by zicklag** [comment]
Other people can probably answer better, but I believe the choice of Preact has something to do with it's smallness and it's ability to operate in the isolated islands that Fresh is built on.

When using React, I think you often need the compiler, and the build toochain and all of this extra _tooling_ just to use it, but with Preact you can insert it into just one component on you page to get some client-side reactivity. So it fits well for both the island and the no-build-step nature of Fresh.

Related ( short ) discussion: https://github.com/denoland/fresh/discussions/598.

Also, from what little I've seen, Preact is extremely compatible with many react components, so it kind of inherits a lot of useful ecosystem compatibility from React.

**Comment by lucacasonato** [comment]
Because Preact is excellent. It strikes a great balance between size, usability, simplicity, familiarity and customisability.

**Comment by rschristian** [comment]
> From what little I know about Preact, it is not particularly widely used and its market share doesn't seem to be growing either

Out of curiosity, what makes you say this? A particular write up or survey?

While NPM download counts are hardly the most scientific of values, Preact is solidly in fourth when it comes to JS framework usage (behind React, Vue, and Angular), and has been growing pretty quickly (up 2.25x for the year). I totally get that, even if we say growing downloads == growing user numbers, that doesn't necessarily mean a growth in market share, but it should be more than popular enough to keep its ecosystem going.

**Reply by FeldrinH** [reply]
Partly various surveys, where [State of JS 2021](https://2021.stateofjs.com/en-US/libraries/front-end-frameworks) shows Preact having very minimal growth over the last couple of years and [State of Frontend 2022](https://tsh.io/state-of-frontend/) and [Stack Overflow Developer Survey 2022](https://survey.stackoverflow.co/2022/#section-most-popular-technologies-web-frameworks-and-technologies) have not included Preact at all.

Partly the fact that of the handful of webdev projects I've been a part of and know of, I've never seen Preact used or event discussed as a possible choice.

Admittedly none of this is strong evidence of anything, but for an outsider it generally gives the impression that Preact isn't seeing much adoption.

**Reply by rschristian** [reply]
Thanks!

With the surveys there's some discomfort with recommending/telling users to go fill them in, and I think Preact is in the slump between "early fervent evangelists" and "has influencers in the ecosystem instructing to do what maintainers might not themselves". Just my experience though.

FWIW, as of the start of the year, [Preact could be found on 4% of websites](https://twitter.com/_developit/status/1483124212332126211). There's a non-zero % chance you're using it without realizing.

**Reply by TranquilMarmot** [reply]
For what it's worth, I thought Preact was more of a 1:1 replacement for React and that's why it's never shown on surveys- they would basically be listing the same thing twice.

Trying out fresh, I definitely see that Preact is quite a bit different from React. 

I'd love to see a world where [`preact-compat`](https://preactjs.com/guide/v10/switching-to-preact/) and `npm`-based React libraries are easier to use with Fresh, since that would unblock some pretty huge roadblocks to building significantly sized apps.

**Reply by vicary** [reply]
I am a big fan of the new paradigm and the performance it brings, deno deploy also plays a big role in the beautiful performance!

I tried porting quite a few of my hobby projects into fresh, and honestly found the ecosystem is no match unless fresh brings preact to the light.

Many commonly used *modules* has a hard dependency on React, even loaded with `skypack` and/or `esm.sh`. For example, MUI and Formik. To resolve this issue, either these popular libraries need to include preact because enough voices are made, or the hosting platforms above allows local import maps to forcibly substitute their react dependencies similar to `"resolutions"` field in yarn's `package.json`. i.e. `{ "react": "https://esm.sh/preact/compat" }`

**Reply by rschristian** [reply]
@vicary What do you mean by "brings preact into the light"? Again, it's not like it's a niche project.

`esm.sh` also has aliasing built-in, see: https://github.com/denoland/fresh/discussions/432

**Reply by vicary** [reply]
@rschristian Thanks for sharing. While the alias is a nice touch, dependencies are usually more than that. For example `Formik` expects a `SyntheticEvent`/`FormEvent` AND `ReactElement` as children, while preact works on native DOM events and has it own `VNode` interface. This is only one example on how far the ecosystem is away from a trivial alias.

No matter how popular we want preact to be, it is just not (yet) the default.

I see that `pnpm` is actually gaining support for many packages, I hope preact will have it's place in the market just like `pnpm`.

**Comment by katopz** [comment]
Let's the data say for itself https://krausest.github.io/js-framework-benchmark/current.html
<img width="668" alt="image" src="https://user-images.githubusercontent.com/97060/199471583-0bce0205-c3c0-4694-ba61-f8bdf1d4bff4.png">

Anyway, I did try `Svetle` but I don't feel right for some reason.

**Comment by RussellPekko** [comment]
I know preact is excellent, but what if i just want to use native react ? Some react-based libs such as framer-motion could not run well in preact. It would be better if users could freely choose react or preact

**Reply by zicklag** [reply]
I'm pretty sure it's just not practical for Fresh, with it's island model and zero-compilation, to work with React. I could be wrong, but I believe it's not just preference, but the fact that React's design is fundamentally not compatible with the design decisions of Fresh.

**Reply by RussellPekko** [reply]
Since deno can support npm, I think it is not difficult for fresh to add support for react, which can also bring a considerable number of users to fresh

**Reply by rschristian** [reply]
Supporting NPM really really isn't a barrier/unblocker in any way. You could always compile and toss React up on a CDN if you wanted to.

The bigger issue is that it really wouldn't make much sense to use React in this way. Loading ~45kb gz of JS just to hydrate parts of the page makes little sense. The cost is too great.

Not to mention, the React team doesn't want users using React in this way really; they're sold on RSCs being the new direction and partial hydration isn't of interest. That adds quite a bit of fragility into tools providing that.

**Reply by AddictArts** [reply]
> I'm pretty sure it's just not practical for Fresh, with it's island model and zero-compilation, to work with React. I could be wrong, but I believe it's not just preference, but the fact that React's design is fundamentally not compatible with the design decisions of Fresh.

However, a person may not care about the island model for the react part of a page or route; Therefore, it is not really a question of fresh react integration, more fresh is just a webserver and let the browser do the react thing. Sure, not ideal for some use cases may help with bringing fresh into the picture if you have a way to keep the react investment.

**Reply by AddictArts** [reply]
> Supporting NPM really really isn't a barrier/unblocker in any way. You could always compile and toss React up on a CDN if you wanted to.
> 
> The bigger issue is that it really wouldn't make much sense to use React in this way. Loading ~45kb gz of JS just to hydrate parts of the page makes little sense. The cost is too great.
> 
> Not to mention, the React team doesn't want users using React in this way really; they're sold on RSCs being the new direction and partial hydration isn't of interest. That adds quite a bit of fragility into tools providing that.

I disagree, not an issue Fresh should care about. There are use cases where it would be nice to consider Fresh and leverage an existing react code base. Also, why should Fresh care? Let the people do what they want if it makes sense for them. Sure it is not ideal, but if you can deploy fresh now and have some piece of your website be react then that is a win for Fresh. It doesn't hurt Fresh to be open to client side frameworks.

**Reply by rschristian** [reply]
> However, a person may not care about the island model for the react part of a page or route

To put it simply, if you don't care about Fresh's main selling point, Fresh probably isn't the best choice for you. And that's fine, there are other tools out there that will make you much happier.

> I disagree, not an issue Fresh should care about.

It doesn't look like you've maintained anything in the way of OSS, so you probably can't appreciate the workload that going off the beaten path brings. The React team have been quite open with their dislike of tools like Astro and preference of Next, so if you're creating a tool in the space, you need to be aware of what the future looks like. For non-RSC islands, it looks pretty bleak w/ React.

> It doesn't hurt Fresh to be open to client side frameworks.

Another instance of not being able to appreciate the workload if you haven't done OSS: it massively increases the maintenance burden to support multiple frameworks, and also limits features greatly, as you're limited by the least common denominator.

**Reply by AddictArts** [reply]
@rschristian I am tempted to be snarky, but I digress. We don't know each other and any GitHub searches or other don't really let you know anything about me. Regardless of what I do or have done a different point of view shouldn't be shot down by guessing about me personally.

**Reply by rschristian** [reply]
I have no power to shoot down anything here, just to be clear. I'm unaffiliated with Deno and Fresh, at best I might've made a docs PR once?

What you do or have done certainly affects your outlook on the viability of different features, and that's why I was hinting your (appearance of) lack of experience here isn't helping you appreciate the complexity. I have maintained OSS build tooling for years, and users always want to expand support, it's a constant ask. But it's difficult enough to support a singular external library alongside the features that really defined the tool -- adding other frameworks on top can quickly become a nightmare. It's a really hard thing to understand until you've been there, I certainly struggled with it.

It would hurt Fresh as what you suggest is to water down the identity ("a person may not care about the island model") and take resources away from some features in order to figure out React compat; not to mention, all future work also need to then support React too. It's always a balancing act. 

For what it's worth, [Aleph](https://github.com/alephjs/aleph.js) might be closer to what you're looking for from the sounds of it.

**Reply by AddictArts** [reply]
@rschristian et all.

I didn't suggest, never did .... I am not suggesting Fresh do anything, don't know why that was assumed. Just making the observation that the client, **not fresh server** can run any **js** it wants. Its a browser and it is not owned or controlled by fresh. Sure they miss out on features, no integration like the preact story. This does not hurt fresh in anyway.

**Reply by kylesloper** [reply]
Totally with @AddictArts here. A framework is pretty personal to the developer. To expect to move away from that to use the features Fresh provides is beyond unrealistic and unreasonable. Whilst it does increase complexity, allowing custom client-side frameworks far outweighs the bad in my opinion. I hope a system a little less opinionated (client-side) is considered by the team.

**Comment by ooker777** [comment]
Quoting @marvinhagemeister from [discord discussion](https://discord.com/channels/684898665143206084/991511118524715139/1194963261708304435):

> Lots of frameworks support TypeScript or could be used as underpinnings for the island architecture (see Astro). Fresh mainly picked Preact because it makes it easy to integrate with because it's pretty much the only framework that has a pluggable API to hook into the rendering mechanism at the time. For all other Frameworks you'd need to enhance the framework itself to support islands, but with Preact you can add it without having to change Preact itself. The fact that it's small and has a familiar API for many web developers helps too.
> 
> What Fresh gives you compared to using Preact alone is:
> 
> - routing automatic
> - bundling of islands and the machinery to make them interactive on page load (remember Preact is just a JS lib not a build tool)
> - a serialization layer that supports preact signals to transfer island props to the browser
> - an API layer
> - dev stuff like reloading the page when a component changes
> - a server in general
> 
> All in all Fresh is a somewhat minimal wrapper around Preact. It's a similar relationship as Next.js has with React.

Two talks of @lucacasonato about Fresh:
- [Fresh: a new full stack web framework for Deno with Luca Casonato - YouTube](https://www.youtube.com/watch?v=nBrcmlrekV4 "Fresh: a new full stack web framework for Deno with Luca Casonato - YouTube") on Nick Taylor channel. This one focuses on introducing Fresh
- [Deno's Fresh new web framework with Luca Casonato (JS Party Live!) - YouTube](https://www.youtube.com/watch?v=8Yypl20Dfl4 "Deno's Fresh new web framework with Luca Casonato (JS Party Live!) - YouTube") on Changelog channel. This one deeps dive into design rationale I guess

I also make this as a PR: https://github.com/denoland/fresh/pull/2210/commits/b7408a3208d003cc98f4312369ec5a4f6bfc3b34
